### ISSUE TITLE
Assert if exported FTN message is re-imported

### ISSUE THREAD

**Issue by NuSkooler** [NuSkooler/enigma-bbs#59]
**Problem**:
If a message(s) are exported via FTN a MSGID is generated. If a faulty node sends any of these messages back for import, an assert will be triggered and the system will crash.

This is due to the messages UUIDs for exported vs imported being different and therefor passing duplicate checks: The original UUID is generated while the imported UUID is a named uuid generated from MSGID and other data _at the time of export_.

**Expected behavior**
* Do not crash!
* We should probably have a duplicate message area for sysops only by default where duplicate message(s) land

**Possible Solution**:
Check for existing MSGID directly during dupe check.

**Comment by apamment** [comment]
I'm not sure if this is related, but during the fsxnet flood where a ton of messages were being reexported, ftn_bso.js would fail an assertion on line 787. I'm not really sure what the correct thing to do here is.

**Comment by tracker1** [comment]
Should probably create an MD5 hash of the message content with signature/tags/system/tear lines removed... and use that for part of dupe checking... Also, should check the date/from/to/subject fields for potential conflicts.

**Comment by NuSkooler** [comment]
@apamment Yes this is the same issue. For now the assert can be commented out. This will result in dupes in your message bases, but I think you'll avoid any crashing. The real fix is some better dupe handling and somewhere to put said dupes (both covered in this bug)

@tracker1 A SHA-1 based system is already in place for cases in which MSGID is not present, see [ftn_util.js, Line 124](https://github.com/NuSkooler/enigma-bbs/blob/master/core/ftn_util.js#L124). This is not the problem here (see original description).

Main issue is external message bases in enig are treated as just that: external. FTN MSGIDs are only generated when/if a message is exported; **It's message UUID was generated prior to this** (random), before a MSGID/etc. is established. If the _same_ message is re-imported, a UUID is generated (named/sha-1) -- but it will not match the first one.

Various options (there may be more) I'll be exploring:

1. Create a meta value @ export to hold the UUID generated from MSGID/area; Check this UUID @ import.
2. Inspect MSGID directly for duplicates @ import
3. Same as #1, but use alternate UUID generation (mentioned above in `ftn_util.js`). This might be the best option.

Priority probably needs bumped on this as it's causing real world issues.

**Comment by NuSkooler** [comment]
I looked into this tonight & believe this is now resolved as of f87e9917a09b5d400cb9302ca225611cf16bb979 and 3bbdd4abae2601b551e58a0345a861d864b47bf3.

The fix involves a new way of generating UUIDs similar to (3) mentioned in the comment above. Additionally, MSGID kludges are checked directly if present.

While looking into this, I also noticed I hadn't resolved a TODO item in `database.js` during table creation. The table `message_meta` did not have a `ON DELETE CASCADE` attached to the FK in message. Therefore, any time a message trim occurred, message_meta was not cleaned up. **This could result in inconsistent data** for one, and a ton of wasted storage/entries.

**Due to the above, I suggest the following**:

1. Stop your board
2. Back up db/message.sqlite
3. Create the new table layout for message_meta:

```
sqlite3 db/message.sqlite
ALTER TABLE message_meta RENAME TO message_meta_backup;
CREATE TABLE IF NOT EXISTS message_meta (
message_id INTEGER NOT NULL,
meta_category INTEGER NOT NULL,
meta_name VARCHAR NOT NULL,
meta_value VARCHAR NOT NULL,
UNIQUE(message_id, meta_category, meta_name, meta_value), 
FOREIGN KEY(message_id) REFERENCES message(message_id) ON DELETE CASCADE
);
INSERT INTO message_meta SELECT * FROM message_meta_backup;
```

The mentioned changes are pushed to Xibalba and seem to do the trick. I am going to close this one and create a new issue for a duplicate message area (e.g. drop dupes there vs ignoring them as is the current workings)
